### PR TITLE
[FLINK-2627][utils]Make Scala Data Set utils easier to access

### PR DESCRIPTION
**Pull request by sachingoel0101**
Introduces a package object for Scala data set utils to simplify usage.
New usage: `import org.apache.flink.api.scala.utils._`

### CODE REVIEW

File: flink-scala/src/main/scala/org/apache/flink/api/scala/utils/package.scala [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+  implicit class DataSetUtils[T](val self: DataSet[T]) {
+
+    /**
+     * Method that takes a set of subtask index, total number of elements mappings
+     * and assigns ids to all the elements from the input data set.
+     *
+     * @return a data set of tuple 2 consisting of consecutive ids and initial values.
+     */
+    def zipWithIndex(implicit ti: TypeInformation[(Long, T)],
```
Comment (StephanEwen) [inline flink-scala/src/main/scala/org/apache/flink/api/scala/utils/package.scala]: Quick question: In most other parts of the Scala API, the TypeInformation is passed via context bounds. Even though that de-sugars to an implicit parameter, why not keep the style consistent over all functions?

Comment (sachingoel0101) [inline flink-scala/src/main/scala/org/apache/flink/api/scala/utils/package.scala]: I'm not sure I understand. I'm not familiar with implicit values and type information systems of scala as well as flink.

Comment (tillrohrmann) [inline flink-scala/src/main/scala/org/apache/flink/api/scala/utils/package.scala]: Stephan suggested to remove the implicit parameter lists from all methods and write instead `implicit class DataSetUtils[T: TypeInformation: ClassTag](val self: DataSet[T])`. +1 for his suggestion.

**Comment by sachingoel0101** [comment]
I am unable to get rid of the implicit type information for the `zip` functions, presumably because the type information for `(Long,T)` isn't found.

**Comment by tillrohrmann** [comment]
Have you tried constructing the type information explicitly `createTypeInformation[(Long, T)]`?

**Comment by sachingoel0101** [comment]
I get the following error on using that: 
`macro implementation not found: createTypeInformation (the most common reason is that you cannot use macro implementations in the same compilation run that defines them)`
The correct place to use that would be just before the `wrap` call though, right?

**Comment by StephanEwen** [comment]
I think for this, the type information should be passed from the call site, you should not need to create it explicitly.

**Comment by StephanEwen** [comment]
Ah, sorry, my bad, you should take the TypeInformation fro `T` from the call site.

You may need to manually create the type info for the tuple, from the `T` type info, by creating a case class type info for `Tuple2` with `Long` and `T`.

**Comment by tillrohrmann** [comment]
To be precise, directly before the `map` call. And you have to make declare the value as an implicit value. Otherwise, the map call won't find it.

File: flink-scala/src/main/scala/org/apache/flink/api/scala/utils/package.scala [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+   * This class provides simple utility methods for zipping elements in a data set with an index
+   * or with a unique identifier, sampling elements from a data set.
+   *
+   * @param self Data Set
+   */
+
+  implicit class DataSetUtils[T: TypeInformation: ClassTag](val self: DataSet[T]) {
+
+    implicit val tupleTypeInformation = new CaseClassTypeInfo[(Long, T)](
+      classOf[(Long, T)],
```
Comment (sachingoel0101) [inline flink-scala/src/main/scala/org/apache/flink/api/scala/utils/package.scala]: @StephanEwen , is this what you had in mind? 
Thanks a lot. Figuring this out cleared up a lot of things for me. :)

Comment (tillrohrmann) [inline flink-scala/src/main/scala/org/apache/flink/api/scala/utils/package.scala]: Didn't `createTypeInformation[(Long, T)]` work?

Comment (sachingoel0101) [inline flink-scala/src/main/scala/org/apache/flink/api/scala/utils/package.scala]: No. It always led to error I mentioned above.

**Comment by StephanEwen** [comment]
I think the way to go is not to put the type info into the class, but into the methods, and create it as follows:

```scala
def zipWithIndex[T : TypeInformation : ClassTag](): DataSet[(Long, T)] = {
  val tInfo = implicitly[TypeInformation[T]]
  
  implicit val tupleTypeInformation = new CaseClassTypeInfo[(Long, T)](
      classOf[(Long, T)],
      Array(BasicTypeInfo.LONG_TYPE_INFO, tInfo),
      Seq(BasicTypeInfo.LONG_TYPE_INFO, tInfo),
      Array("_1", "_2"))

  wrap(jutils.zipWithIndex(self.javaSet)).map { t => (t.f0.toLong, t.f1) }
```

All the methods in the utils class should have parenthesis, they are not a side effect free getters after all.

Also, some tooling around creating Scala Tuple type information would be nice. I can see that there are more places where one would do that.

**Comment by sachingoel0101** [comment]
@StephanEwen , I have created a separate function to create type information for 2-tuple.

One question though. Why is there a need to generate type information explicitly here? The `TypeInformationGen` class does have a case analysis for `Product` types. I may be very wrong, but `createTypeInformation` macro cannot be used anywhere inside the module itself but only after the module's been compiled. This is perhaps why `createTypeInformation` works in, say, `flink-ml`.

**Comment by StephanEwen** [comment]
The type information generation works as a macro on the abstract syntax tree, that's why it cannot work on its own code (or any code in the same project, which is the same compilation unit).

File: flink-scala/src/main/scala/org/apache/flink/api/scala/package.scala [hunk 3]
```diff
@@ -70,4 +73,27 @@ package object scala {
     }
     st(depth).toString
   }
+
+  def createTuple2TypeInformation[T1, T2]
+  (implicit t1: TypeInformation[T1], t2: TypeInformation[T2])
```
Comment (StephanEwen) [inline flink-scala/src/main/scala/org/apache/flink/api/scala/package.scala]: Making this implicit seems dangerous, it should be explicitly provided.

File: flink-scala/src/main/scala/org/apache/flink/api/scala/UnfinishedCoGroupOperation.scala [hunk 4]
```diff
@@ -61,33 +59,14 @@ class UnfinishedCoGroupOperation[L: ClassTag, R: ClassTag](
 
     // We have to use this hack, for some reason classOf[Array[T]] does not work.
     // Maybe because ObjectArrayTypeInfo does not accept the Scala Array as an array class.
-    val leftArrayType =
+    implicit val leftArrayType =
```
Comment (StephanEwen) [inline flink-scala/src/main/scala/org/apache/flink/api/scala/UnfinishedCoGroupOperation.scala]: These values should be explicitly provided, not implicitly. It makes the code much more understandable

**Comment by StephanEwen** [comment]
I think this looks good.

I personally like explicit arguments, even if it means sometimes explicitly summoning an implicit bound. It just makes it clearer where the type infos flow, which is not that easy to figure out ;-)

This is subject to debate, though. @tillrohrmann and @aljoscha may be of different opinion here...

**Comment by sachingoel0101** [comment]
I think I agree with that. I wasn't too happy about using implicit arguments here; we're constructing the type information explicitly anyway. 
Will push a commit in a while to change this.

**Comment by aljoscha** [comment]
I took a stab at removing the implicit parameters here: https://github.com/aljoscha/flink/commit/e197ea4aba4005400bc80a5693f17ec2617bfae5
The tests are still running on Travis but I think it should work.

**Comment by sachingoel0101** [comment]
Ah. Thank you @aljoscha. Travis should pass. I've already pushed a fix.

**Comment by sachingoel0101** [comment]
Travis passes successfully. I've squashed the commits.
This should be mergeable now.

**Comment by sachingoel0101** [comment]
@StephanEwen , can you look this over again?

**Comment by sachingoel0101** [comment]
Hey @StephanEwen, apologies for being too eager but is it possible to get this in soon?

**Comment by sachingoel0101** [comment]
Unrelated failures. Already filed jiras for those. 2700 and 2612.

**Comment by StephanEwen** [comment]
Looks good to me.

+1 to merge

File: flink-scala/src/main/scala/org/apache/flink/api/scala/utils/package.scala [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+package object utils {
+
+  /**
+   * This class provides simple utility methods for zipping elements in a data set with an index
+   * or with a unique identifier, sampling elements from a data set.
+   *
+   * @param self Data Set
+   */
+
```
Comment (tillrohrmann) [inline flink-scala/src/main/scala/org/apache/flink/api/scala/utils/package.scala]: line break

**Comment by tillrohrmann** [comment]
LGTM.

+1 for merging :-)

**Comment by StephanEwen** [comment]
Will address Till's comment and merge this...

**Comment by sachingoel0101** [comment]
I've already removed the line break. :)

**Pull request closed without merging**